Integration with main website

Ampretuzo opened the pull request:
WIP

rschroll commented on app/client/src/features/progress/components/SideNav.js:
```diff
@@ -84,8 +115,20 @@ class SideNav extends Component<Props> {
+            <TmpButton style={{ margin: '10px' }} onClick={ this.handleResetFormToSavedStateClick }>
+              Reset to Saved Values
+            </TmpButton>
+            <TmpButton style={{ margin: '10px' }} onClick={ this.handleUploadPdfClick }>
```
Can we disable this button when uploads are not allowed?

Ampretuzo replied:
Sure - done!

rschroll commented on app/client/src/features/tdi/actions.js:
```diff
@@ -0,0 +1,170 @@
+        body: data
+      }
+
+      try {
```
Can we also POST the JSON to the website here?  (Unless that is happening and I missed it.)

Ampretuzo replied:
Posting the json just in case would not hurt, but pretty much every reasonable situation is covered by the fact that saved values must correspond to displayed document before publishing.
(Except probably when an admin and the Fellow do concurrent edits - I made a deliberate choice not to support concurrency, even a non-blocking one.)

rschroll commented on app/client/src/features/tdi/actions.js:
```diff
@@ -0,0 +1,170 @@
+export function initializeApplication(history: RouterHistory, fellowKeyUrlsafe: ?string): AsyncAction {
+  return async (dispatch, getState) => {
+    if (!fellowKeyUrlsafe) { // Not an admin.
+      return
```
If there's no saved state, can we trigger a download of the JSON and redirect to the editor?

Ampretuzo replied:
That should be working now. :crossed_fingers: 
Was a bit tricky to set up actually - we have to wait until the form is mounted until we can call reset on it. Also, form initialization is contending with json upload functionality.

rschroll commented on app/client/src/features/tdi/actions.js:
```diff
@@ -0,0 +1,170 @@
+// NOTE: I'd rather duplicate fetchFellowData while a strong similarity becomes
+// apparent. I'm mainly trying to keep some headroom for diverging error-handling
+// in case of admins.
+function fetchFellowDataWithKey(fellowKeyUrlsafe: string): AsyncAction {
```
I haven't considered closely, but would it be easier to enforce that everyone, regular users and admins, come via a key?  That would decrease duplication and decrease the number of corner cases to fall in.

Ampretuzo replied:
That was my intention in the beginning too, but:
1. it was easier to bootstrap the app during development without worrying about keys. (Just a historical reason.)
2. And in case someone decides to remember the url by heart, `/resumake/` is definitely much easier than `/resumake/<urlsafe_fellow_key>`.
The admins have no choice but to use the url with the key from the website, so that the app can initialize.

Ampretuzo committed:
rm TmpButton.js

Ampretuzo committed:
Button "disabled" styling

Ampretuzo committed:
Disable publish button

Ampretuzo committed:
Install react-tooltip

NOTE: I'm not a fan of adding dependencies for such simple use cases - we should probably remove the tooltip.

Ampretuzo committed:
Tooltip and help text when Publish is disabled

Ampretuzo committed:
Single fetch function

Ampretuzo committed:
Initial fellow data load

Ampretuzo committed:
Store the key on initial admin load

Ampretuzo committed:
Less suggestive publish help wording

Ampretuzo commented:
@rschroll I'll re-route this PR to your branch, since you already merged this in today.
Even though what we're trying to accomplish is simple enough, I think the complexity quickly gets out of hand: look at one flag, write another, that kind of stuff.
Especially when developing against the dev server - without tests.
We should probably refactor everything and put some tests on it.
